### PR TITLE
storage: remove meta, save commit log in CF_WRITE.

### PR DESCRIPTION
**Pull request by disksing**
@siddontang 
@zhangjinpeng1987 
@BusyJay 
@ngaut

### CODE REVIEW

**Comment by siddontang** [comment]
do you delete all write cf data before applying snapshot?

File: src/storage/mvcc/lock.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        match *mutation {
+            Mutation::Put(_) => LockType::Put,
+            Mutation::Delete(_) => LockType::Delete,
+            Mutation::Lock(_) => LockType::Lock,
+        }
+    }
+
+    fn from_u8(b: u8) -> Option<LockType> {
+        match b {
+            b'P' => Some(LockType::Put),
```
Comment (siddontang) [inline src/storage/mvcc/lock.rs]: should define a const

**Comment by disksing** [comment]
Yes. It will be deleted. In fact, all CFs are iterated. See https://github.com/pingcap/tikv/pull/907/files#diff-ac5fa05712c478a85585b4852858d036R443

**Comment by siddontang** [comment]
got it

File: src/storage/mvcc/mod.rs [hunk 2]
```diff
@@ -40,6 +42,7 @@ quick_error! {
             description("key is locked (backoff or cleanup)")
             display("key is locked (backoff or cleanup) {}-{}@{}", escape(key), escape(primary), ts)
         }
+        BadFormatLock {description("bad format lock data")}
         AlreadyCommitted {commit_ts: u64} {
```
Comment (siddontang) [inline src/storage/mvcc/mod.rs]: Committed is ok ?

**Comment by siddontang** [comment]
any benchmark?

**Comment by siddontang** [comment]
@zhangjinpeng1987 could you help @disksing to do benchmark?

File: src/storage/mvcc/lock.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    pub fn new(lock_type: LockType, primary: Vec<u8>, ts: u64) -> Lock {
+        Lock {
+            lock_type: lock_type,
+            primary: primary,
+            ts: ts,
+        }
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut b = vec![];
```
Comment (siddontang) [inline src/storage/mvcc/lock.rs]: pre-alloc?

File: src/storage/mvcc/reader.rs [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    pub fn reverse_seek_write(&mut self, key: &Key, ts: u64) -> Result<Option<(u64, u64)>> {
+        self.seek_write_impl(key, ts, true)
+    }
+
+    fn seek_write_impl(&mut self, key: &Key, ts: u64, reverse: bool) -> Result<Option<(u64, u64)>> {
+        let (k, v) = {
+            if self.write_cursor.is_none() {
+                self.write_cursor = Some(try!(self.snapshot.iter_cf(CF_WRITE)));
+            }
+            let mut cursor = self.write_cursor.as_mut().unwrap().as_mut();
```
Comment (BusyJay) [inline src/storage/mvcc/reader.rs]: Why `as_mut` twice?

File: src/storage/txn/scheduler.rs [hunk 5]
```diff
@@ -238,60 +236,50 @@ fn process_read(cid: u64, cmd: Command, ch: SendCh<Msg>, snapshot: Box<Snapshot>
...
-                        lock.set_primary_lock(pb.take_primary_key());
-                        lock.set_lock_version(pb.get_start_ts());
-                        lock.set_key(try!(key.raw()));
-                        locks.push(lock);
+            let mut reader = MvccReader::new(snapshot.as_ref());
+            let res = reader.scan_lock(|lock| lock.ts <= max_ts)
+                .map_err(Error::from)
+                .and_then(|v| {
+                    let mut locks = vec![];
+                    for (key, lock) in v.into_iter() {
```
Comment (BusyJay) [inline src/storage/txn/scheduler.rs]: `into_iter` is unnecessary.

File: src/storage/mvcc/reader.rs [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    pub fn load_lock(&mut self, key: &Key) -> Result<Option<Lock>> {
+        if let Some(ref mut cursor) = self.lock_cursor {
+            match try!(cursor.get(&key)) {
+                Some(v) => Ok(Some(try!(Lock::parse(v)))),
+                None => Ok(None),
+            }
+        } else {
+            match try!(self.snapshot.get_cf("lock", &key)) {
```
Comment (siddontang) [inline src/storage/mvcc/reader.rs]: use const name here

File: src/storage/mvcc/reader.rs [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+                let mut cursor = self.data_cursor.as_mut().unwrap();
+                if !try!(cursor.seek(&key)) {
+                    return Ok(None);
+                }
+                try!(Key::from_encoded(cursor.key().to_vec()).truncate_ts())
+            };
+            if let Some(v) = try!(self.get(&key, ts)) {
+                return Ok(Some((key, v)));
+            }
+            key = key.append_ts(0);
```
Comment (siddontang) [inline src/storage/mvcc/reader.rs]: why appending 0 here?

Comment (disksing) [inline src/storage/mvcc/reader.rs]: To make sure next `seek` will return next key instead of next version of current key.

Comment (siddontang) [inline src/storage/mvcc/reader.rs]: so why reverse_seek seems return previous version of the current key, not previous key?

Comment (disksing) [inline src/storage/mvcc/reader.rs]: The key without version should always be in the first place, so `reverse_seek(key)` will bring us to previous key correctly.

**Comment by ngaut** [comment]
Much simpler.

**Comment by siddontang** [comment]
LGTM

PTAL @hhkbp2

File: src/storage/mvcc/reader.rs [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+            }
+        }
+        match try!(self.seek_write(key, ts)) {
+            Some((start_ts, _)) => self.load_data(key, start_ts),
+            None => Ok(None),
+        }
+    }
+
+    pub fn get_txn_commit_ts(&mut self, key: &Key, start_ts: u64) -> Result<Option<u64>> {
+        match try!(self.reverse_seek_write(key, start_ts)) {
```
Comment (BusyJay) [inline src/storage/mvcc/reader.rs]: Why `reverse_seek` here?

File: src/storage/mvcc/reader.rs [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    fn create_data_cursor(&mut self) -> Result<()> {
+        if self.data_cursor.is_none() {
+            self.data_cursor = Some(try!(self.snapshot.iter()));
+        }
+        Ok(())
+    }
+
+    pub fn seek(&mut self, mut key: Key, ts: u64) -> Result<Option<(Key, Value)>> {
```
Comment (BusyJay) [inline src/storage/mvcc/reader.rs]: What does `ts` mean here?

File: src/storage/mvcc/reader.rs [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    fn seek_write_impl(&mut self, key: &Key, ts: u64, reverse: bool) -> Result<Option<(u64, u64)>> {
+        let (k, v) = {
+            if self.write_cursor.is_none() {
+                self.write_cursor = Some(try!(self.snapshot.iter_cf(CF_WRITE)));
+            }
+            let mut cursor = self.write_cursor.as_mut().unwrap();
+            let ok = if reverse {
+                try!(cursor.reverse_seek(&key.append_ts(ts)))
+            } else {
+                try!(cursor.seek(&key.append_ts(ts)))
```
Comment (BusyJay) [inline src/storage/mvcc/reader.rs]: Should use `near_seek` instead.

File: src/storage/mvcc/reader.rs [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+            };
+            if !ok {
+                return Ok(None);
+            }
+            (cursor.key().to_vec(), cursor.value().to_vec())
```
Comment (BusyJay) [inline src/storage/mvcc/reader.rs]: Maybe we can check key before clone value.

File: src/storage/mvcc/txn.rs [hunk 12]
```diff
@@ -109,63 +87,39 @@ impl<'a> MvccTxn<'a> {
...
-        try!(self.commit_impl(key, commit_ts, &mut meta));
-        self.write_meta(key, &mut meta);
-        Ok(())
-    }
-
-    fn commit_impl(&mut self, key: &Key, commit_ts: u64, meta: &mut Meta) -> Result<()> {
-        let lock_type = match try!(self.snapshot.load_lock(key)) {
-            Some(ref lock) if lock.get_start_ts() == self.start_ts => lock.get_field_type(),
+        let lock_type = match try!(self.reader.load_lock(key)) {
+            Some(ref lock) if lock.ts == self.start_ts => lock.lock_type,
```
Comment (BusyJay) [inline src/storage/mvcc/txn.rs]: Why `ref`? I prefer `assert_eq!(lock.ts, self.start_ts)`.

Comment (BusyJay) [inline src/storage/mvcc/txn.rs]: Hmm, seems `lock.ts` doesn't have to be equal to `self.start_ts`.

File: src/storage/mvcc/reader.rs (hunk 8 above)
Comment (disksing) [inline src/storage/mvcc/reader.rs]: For a committed transaction, its `commit_ts` is always greater than `start_ts`, and no other transaction's `commit_ts` could be between them. So `reverse_seek(start_ts)` on write CF will return the txn's commit entry (if it exists).

File: src/storage/mvcc/reader.rs (hunk 10 above)
Comment (disksing) [inline src/storage/mvcc/reader.rs]: Right. It's a mistake.

File: src/storage/mvcc/reader.rs (hunk 9 above)
Comment (disksing) [inline src/storage/mvcc/reader.rs]: `seek()` should only return values that are written before `ts`.

Comment (disksing) [inline src/storage/mvcc/reader.rs]: `seek()` should only return values that are written before `ts`.

File: src/storage/mvcc/txn.rs (hunk 12 above)
Comment (disksing) [inline src/storage/mvcc/txn.rs]: It won't compile without `ref`.

**Comment by disksing** [comment]
PTAL @BusyJay

File: src/storage/mvcc/lock.rs [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    pub fn new(lock_type: LockType, primary: Vec<u8>, ts: u64) -> Lock {
+        Lock {
+            lock_type: lock_type,
+            primary: primary,
+            ts: ts,
+        }
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut b = Vec::<u8>::with_capacity(1 + MAX_VAR_U64_LEN + self.primary.len() +
```
Comment (BusyJay) [inline src/storage/mvcc/lock.rs]: `<u8>` no need here.

**Comment by BusyJay** [comment]
rest LGTM

**Comment by BusyJay** [comment]
LGTM

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**